File: src/core/numbering.ts
// src/core/numbering.ts
import type { C3Graph } from "./walker";

export function nextContainerId(graph: C3Graph): number {
  const containers = graph.byType.get("container") || [];
  const nums = containers
    .map(c => parseInt(c.id.replace("c3-", ""), 10))
    .filter(n => !isNaN(n));
  return nums.length === 0 ? 1 : Math.max(...nums) + 1;
}

export function nextComponentId(graph: C3Graph, containerNum: number, feature: boolean): string {
  const prefix = `c3-${containerNum}`;
  const components = (graph.byType.get("component") || [])
    .filter(c => c.id.startsWith(prefix))
    .map(c => parseInt(c.id.replace(prefix, ""), 10))
    .filter(n => !isNaN(n));

  if (feature) {
    // Feature: 10+
    const featureNums = components.filter(n => n >= 10);
    const next = featureNums.length === 0 ? 10 : Math.max(...featureNums) + 1;
    return `c3-${containerNum}${String(next).padStart(2, "0")}`;
  } else {
    // Foundation: 01-09
    const foundationNums = components.filter(n => n >= 1 && n <= 9);
    const next = foundationNums.length === 0 ? 1 : Math.max(...foundationNums) + 1;
    if (next > 9) throw new Error(`Container c3-${containerNum} has no more foundation slots (01-09 full)`);
    return `c3-${containerNum}${String(next).padStart(2, "0")}`;
  }
}

export function nextAdrId(slug: string): string {
  const date = new Date().toISOString().slice(0, 10).replace(/-/g, "");
  return `adr-${date}-${slug}`;
}

File: cli/internal/numbering/numbering.go
package numbering

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/lagz0ne/c3-design/cli/internal/frontmatter"
	"github.com/lagz0ne/c3-design/cli/internal/walker"
)

// NextContainerId returns the next available container number (1, 2, 3, ...).
func NextContainerId(graph *walker.C3Graph) int {
	containers := graph.ByType(frontmatter.DocContainer)
	if len(containers) == 0 {
		return 1
	}

	max := 0
	for _, c := range containers {
		numStr := strings.TrimPrefix(c.ID, "c3-")
		n, err := strconv.Atoi(numStr)
		if err != nil {
			continue
		}
		if n > max {
			max = n
		}
	}
	return max + 1
}

// NextComponentId returns the next available component ID for a container.
// Foundation components use slots 01-09, feature components use 10+.
func NextComponentId(graph *walker.C3Graph, containerNum int, feature bool) (string, error) {
	prefix := fmt.Sprintf("c3-%d", containerNum)
	components := graph.ByType(frontmatter.DocComponent)

	var nums []int
	for _, c := range components {
		if !strings.HasPrefix(c.ID, prefix) {
			continue
		}
		numStr := strings.TrimPrefix(c.ID, prefix)
		n, err := strconv.Atoi(numStr)
		if err != nil {
			continue
		}
		nums = append(nums, n)
	}

	if feature {
		max := 9
		for _, n := range nums {
			if n >= 10 && n > max {
				max = n
			}
		}
		next := max + 1
		return fmt.Sprintf("c3-%d%02d", containerNum, next), nil
	}

	// Foundation: 01-09
	max := 0
	for _, n := range nums {
		if n >= 1 && n <= 9 && n > max {
			max = n
		}
	}
	next := max + 1
	if next > 9 {
		return "", fmt.Errorf("container c3-%d has no more foundation slots (01-09 full)", containerNum)
	}
	return fmt.Sprintf("c3-%d%02d", containerNum, next), nil
}

// NextAdrId generates an ADR ID with today's date and the given slug.
func NextAdrId(slug string) string {
	date := time.Now().Format("20060102")
	return fmt.Sprintf("adr-%s-%s", date, slug)
}
